pandas/core/frame.py

```diff
@@ -1664,11 +1664,12 @@ def from_records(
             else:
                 try:
                     index_data = [arrays[arr_columns.get_loc(field)] for field in index]
+                except KeyError:
+                    # raised by get_loc
```

Comment:
Unfortunately, there are cases that `get_loc` returns a different error .. (if you look in the get_loc code, there are eg a few places where we deliberately raise a TypeError. Separate issue, but we should maybe consider changing that) 
I don't know if this in the end will have any impact on the result here (as those cases probably mean you do not have correct input data anyway), but probably something to check to be sure.